libgadget/lenstools.c

```diff
@@ -0,0 +1,412 @@
+    double **array = malloc(Nx * sizeof(double *));
+    if (array == NULL) {
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(EXIT_FAILURE);
+    }
+    for (int i = 0; i < Nx; i++) {
+        array[i] = malloc(Ny * sizeof(double));
+    }
+    return array;
+}
+
+// Function to free a 2D array
+void free_2d_array(double **array, int Nx) {
+    for (int i = 0; i < Nx; i++) {
+        free(array[i]);
+    }
+    free(array);
+}
+
+typedef struct {
+    int nx, ny, nz;
+} GridDimensions;
+
+// Function to determine the bin index for a given value
+int find_bin(double value, double *bins, int num_bins) {
+    for (int i = 0; i < num_bins - 1; i++) {
+        if (value >= bins[i] && value < bins[i + 1]) {
+            return i;
+        }
+    }
+    return -1;  // Return the last bin if the value is on the boundary or beyond
+}
+
+void grid3d_nfw(const struct particle_data * Parts, int num_particles, double **binning, GridDimensions dims, double ***density) {
+    
+    double position[3];
+    // Process each particle
+    for (int p = 0; p < num_particles; p++) {
+
+        // remove offset
+        for(int d = 0; d < 3; d ++) {
+            position[d] = Parts[p].Pos[d] - PartManager->CurrentParticleOffset[d];
+            while(position[d] > PartManager->BoxSize) position[d] -= PartManager->BoxSize;
+            while(position[d] <= 0) position[d] += PartManager->BoxSize;
+        }
+        int ix = find_bin(position[0], binning[0], dims.nx + 1);
+        int iy = find_bin(position[1], binning[1], dims.ny + 1);
+        int iz = find_bin(position[2], binning[2], dims.nz + 1);
+
+        // continue if the particle is outside the grid
+        if (ix == -1 || iy == -1 || iz == -1) {
+            continue;
+        }
+        // Increment the density in the appropriate bin
+        density[ix][iy][iz]++;
+    }
+
+}
+
+void projectDensity(double ***density, double **density_projected, GridDimensions dims, int normal) {
+    // z; x, y
+    // y; z, x
+    // x; y, z
+    int DimNorm = (normal == 0) ? dims.nx : (normal == 1) ? dims.ny : dims.nz;
+    int Dim0 = (normal == 2) ? dims.nx : (normal == 0) ? dims.ny : dims.nx;
+    int Dim1 = (normal == 2) ? dims.ny : (normal == 1) ? dims.nz : dims.nz;
+
+    for (int i = 0; i < Dim0; i++) {
+        for (int j = 0; j < Dim1; j++) {
+            density_projected[i][j] = 0;
+            for (int k = 0; k < DimNorm; k++) {
+                if (normal == 0) {
+                    density_projected[i][j] += density[k][i][j];
+                } else if (normal == 1) {
+                    density_projected[i][j] += density[i][k][j];
+                } else {
+                    density_projected[i][j] += density[i][j][k];
+                }
+            }
+        }
+    }
+}
+
+void calculate_lensing_potential(double **density_projected, int plane_resolution, double bin_resolution_0, double bin_resolution_1, double chi,double smooth, double **lensing_potential) {
```

Comment:
I think this might be a function we already have in the PM code. Let me look...

Reply:
If not we should rewrite it to be mpi-parallel along the lines of petapm. I can have a go at that.

Reply:
Ignore this: the arrays are small enough that we should not worry about it.